### ISSUE TITLE
Display of ICDs is asymmetric

### ISSUE THREAD

**Issue by edwardchapin** [tmtsoftware/icd#25]
It seems as though the ICD view from subsystem A to subsystem B only shows things published by A subscribed to by B. In order to see things published by B and subscribed to by A, you need to click on the subsystem switcher in the middle-top of the web page.

Examples: ICD from NFIRAOS to TCS
- only items published by the RTC (within NFIRAOS) for the TCS are shown.

If you press the switched button:
- only items published by TCS for NFIRAOS (cmNFIRAOS, tcsPk).

**Comment by abrighton** [comment]
I checked in the fix to the icd master branch.

**Comment by abrighton** [comment]
I published a 0.11 release that includes this fix: See https://github.com/tmtsoftware/icd/releases.

**Comment by edwardchapin** [comment]
Thanks - that's an improvement. However,  I think there is still (a perhaps deeper) issue. Sticking to the NFIRAOS - TCS example, select NFIRAOS first, and TCS second. I can see that lgsWfs subscribes to zenithAngle as published by tcsPk. However, expanding zenithAngle in this view does not give me all of the details. I need to switch the sense of the ICD so that TCS is first, and NFIRAOS second so that I can see it is a floating point value ranging from 0 to 90 degrees.

Normally when writing an ICD, the only purpose of having one subsystem listed before the second is the order they appear in the title. Otherwise the "NFIRAOS to TCS" or "TCS to NFIRAOS" ICDs would be identical documents. As it stands I could only get all of the information by generating both of these documents through the web interface.

**Comment by tmtsoftwareadmin** [comment]
Hi Ed,

I see your point about the two being asymmetric, but while it might be incorrect, this was a conscious decision.

When viewing an ICD or API in the UI now, a subscriber entry is shown with the name and description only and a link to the other system.  You click the link to find out more about the item.

The API/ICD is saying this component uses an item published by another component. This is true for events and commands. Note that a subscriber has a “usage” entry to state how the subscribed information is used in the component.  

The Interface for one subsystem/component is that it uses the other subsystem/component information.  Grabbing information out of the second subsystem to put in the ICD does not necessarily seem like the right thing to do.  It makes it seem like (to me) that somehow the first subsystem is in charge of the information from the second.  I realize this might be subtle and even possibly wrong!

I/We should discuss this with Systems Engineering, but to me the two documents/interfaces are not the same/symmetric.  

Kim

> On Feb 22, 2016, at 5:45 PM, Edward Chapin <notifications@github.com> wrote:
> 
> Thanks - that's an improvement. However, I think there is still (a perhaps deeper) issue. Sticking to the NFIRAOS - TCS example, select NFIRAOS first, and TCS second. I can see that lgsWfs subscribes to zenithAngle as published by tcsPk. However, expanding zenithAngle in this view does not give me all of the details. I need to switch the sense of the ICD so that TCS is first, and NFIRAOS second so that I can see it is a floating point value ranging from 0 to 90 degrees.
> 
> Normally when writing an ICD, the only purpose of having one subsystem listed before the second is the order they appear in the title. Otherwise the "NFIRAOS to TCS" or "TCS to NFIRAOS" ICDs would be identical documents. As it stands I could only get all of the information by generating both of these documents through the web interface.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/tmtsoftware/icd/issues/25#issuecomment-187422792>.
> 

________
Kim Gillies
Software Architect
Thirty Meter Telescope Project
kgillies@tmt.org
+1 626 395 1655

**Comment by edwardchapin** [comment]
Hi,

Yes, it's certainly the case that the information is all there - clicking on something for details is reasonable when you're sitting in front of a web browser. When doing real development work, you might argue that these sorts of views even make more sense (generally speaking you are probably working on only one side of the interface!).

Where I'm coming from is the hope/desire that we can use this system to simplify our documentation deliverables (this year!), for which the ICDs are one component. For example, there is a single ICD defining all of the software interfaces between the TCS and IRIS. I would love to remove all of the detailed tables from the word documents and simply say "refer to the attached PDF". I suppose we could attach *two* PDFs (from each subsystem's point of view), but they will have a lot of overlapping information and it might be confusing. We really need this "hard copy" since there is no central icdwebserver instance that is publicly available to, e.g., reviewers (yet), and probably not something CSW is able/willing to set up and maintain at this stage.

To summarize: I think that the API view is what I would, expect as-is (it focusses on the particular subsystem you have selected - telemetry it generates, commands it implements etc. - providing only links to external telemetry that it consumes. It's only the ICD view I would propose changing to include all details from both subsystems.

Certainly let's contact systems engineering as well, but I think this idea more closely matches the established purpose of ICDs in the TMT context (a single place where people can get all of the details to design/test the systems on boths sides of an interface).

thanks,
Ed

**Comment by tmtsoftwareadmin** [comment]
Hi Ed,

> On Feb 23, 2016, at 2:23 PM, Edward Chapin <notifications@github.com> wrote:
> 
> Hi,
> 
> Yes, it's certainly the case that the information is all there - clicking on something for details is reasonable when you're sitting in front of a web browser. When doing real development work, you might argue that these sorts of views even make more sense (generally speaking you are probably working on only one side of the interface!).
> 

> Where I'm coming from is the hope/desire that we can use this system to simplify our documentation deliverables (this year!), for which the ICDs are one component. For example, there is a single ICD defining all of the software interfaces between the TCS and IRIS. I would love to remove all of the detailed tables from the word documents and simply say "refer to the attached PDF”.
> 
Yes, this is the goal!  In fact, SE has agreed that it should be possible to state that a version of the API or ICD is a link to the IDC database, but they have not confirmed this.

I think publishing the API for IRIS and an ICD from IRIS to TCS should be done if we can.

> I suppose we could attach two PDFs (from each subsystem's point of view), but they will have a lot of overlapping information and it might be confusing. We really need this "hard copy" since there is no central icdwebserver instance that is publicly available to, e.g., reviewers (yet), and probably not something CSW is able/willing to set up and maintain at this stage.
> 

> To summarize: I think that the API view is what I would, expect as-is (it focusses on the particular subsystem you have selected - telemetry it generates, commands it implements etc. - providing only links to external telemetry that it consumes. It's only the ICD view I would propose changing to include all details from both subsystems.
> 
Okay this is the real question.  Whether an ICD should include the information from the target subsystem.  I’m sure we can do it.  For me, it’s an issue of what is right.

I want to do whatever I can to have you use this for the review.
> Certainly let's contact systems engineering as well, but I think this idea more closely matches the established purpose of ICDs in the TMT context (a single place where people can get all of the details to design/test the systems on boths sides of an interface).
> 
I think you are right that people have done it this way, but I think they are doing it to say _this is what information I’m expecting_.  Which is sort of where we are since there isn’t a TCS nor a real TCS API.

So for you to even make an IRIS to TCS ICD, we need to get Jimmy to “okay” his TCS model files.  Do you agree?

Kim

> thanks,
> Ed
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/tmtsoftware/icd/issues/25#issuecomment-187851831>.
> 

________
Kim Gillies
Software Architect
Thirty Meter Telescope Project
kgillies@tmt.org
+1 626 395 1655

**Comment by edwardchapin** [comment]
> So for you to even make an IRIS to TCS ICD, we need to get Jimmy to “okay” his TCS model files. Do you agree?

Yes, I agree! Ultimately the person responsible for implementing commands / generating telemetry is in control of their model files. Unlike the use of Word documents (where we need to "pass it back and forth"), the icd-db system helps us clearly delineate areas of responsibility (and we can even work in parallel, in principle).

Ed

**Comment by tmtsoftwareadmin** [comment]
Hi Ed,

I was discussing the issue about symmetry in the ICD documents with H.Buur and she is with you and convinced me.  We are going to talk to SE next week, but I expect we will make the change.  I’ll talk to Allan to see if it’s a big thing.

Kim

> On Feb 23, 2016, at 3:11 PM, Edward Chapin <notifications@github.com> wrote:
> 
> So for you to even make an IRIS to TCS ICD, we need to get Jimmy to “okay” his TCS model files. Do you agree?
> 
> Yes, I agree! Ultimately the person responsible for implementing commands / generating telemetry is in control of their model files. Unlike the use of Word documents (where we need to "pass it back and forth"), the icd-db system helps us clearly delineate areas of responsibility (and we can even work in parallel, in principle).
> 
> Ed
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/tmtsoftware/icd/issues/25#issuecomment-187882619>.
> 

________
Kim Gillies
Software Architect
Thirty Meter Telescope Project
kgillies@tmt.org
+1 626 395 1655

**Comment by abrighton** [comment]
See mails from Ed.

**Comment by abrighton** [comment]
The latest version in the master branch has addressed this issue. The ICD PDFs now display the published items from both subsystems, with links to subscribers. In addition, summary tables at top give an overview.
